lib/std/math/gamma.zig

```diff
@@ -0,0 +1,328 @@
+            den = den / abs + denominator[i];
+        }
+    }
+    return num / den;
+}
+
+// precise sin(pi * x)
+// but not for very small x and integer x, we handle those already
+fn sinpi(comptime T: type, x: T) T {
+    const xmod2 = @mod(x, 2); // [0, 2]
+    const n = (@as(u8, @intFromFloat(4 * xmod2)) + 1) / 2; // {0, 1, 2, 3, 4}
+    const y = xmod2 - 0.5 * @as(T, @floatFromInt(n)); // [-0.25, 0.25]
+    return switch (n) {
+        0, 4 => @sin(std.math.pi * y),
+        1 => @cos(std.math.pi * y),
+        2 => -@sin(std.math.pi * y),
+        3 => -@cos(std.math.pi * y),
+        else => unreachable,
+    };
+}
+
+const expect = std.testing.expect;
+const expectEqual = std.testing.expectEqual;
+const expectApproxEqRel = std.testing.expectApproxEqRel;
+
+test "math.gamma" {
+    inline for (&.{f32, f64}) |T| {
+        const eps = @sqrt(std.math.floatEps(T));
+        try expectApproxEqRel(@as(T, 120), gamma(T, 6), eps);
+        try expectApproxEqRel(@as(T, 362880), gamma(T, 10), eps);
+        try expectApproxEqRel(@as(T, 6402373705728000), gamma(T, 19), eps);
+
+        try expectApproxEqRel(@as(T, 332.7590766955334570), gamma(T, 0.003), eps);
+        try expectApproxEqRel(@as(T, 1.377260301981044573), gamma(T, 0.654), eps);
+        try expectApproxEqRel(@as(T, 1.025393882573518478), gamma(T, 0.959), eps);
+
+        try expectApproxEqRel(@as(T, 7.361898021467681690), gamma(T, 4.16), eps);
+        try expectApproxEqRel(@as(T, 198337.2940287730753), gamma(T, 9.73), eps);
+        try expectApproxEqRel(@as(T, 113718145797241.1666), gamma(T, 17.6), eps);
+
+        try expectApproxEqRel(@as(T, -1.13860211111081424930673), gamma(T, -2.80), eps);
+        try expectApproxEqRel(@as(T, 0.00018573407931875070158), gamma(T, -7.74), eps);
+        try expectApproxEqRel(@as(T, -0.00000001647990903942825), gamma(T, -12.1), eps);
+    }
+}
+
+test "math.gamma.special" {
+    inline for (&.{f32, f64}) |T| {
+        try expect(std.math.isNan(gamma(T, -std.math.nan(T))));
+        try expect(std.math.isNan(gamma(T, std.math.nan(T))));
+        try expect(std.math.isNan(gamma(T, -std.math.inf(T))));
+
+        try expect(std.math.isNan(gamma(T, -4)));
+        try expect(std.math.isNan(gamma(T, -11)));
+        try expect(std.math.isNan(gamma(T, -78)));
+
+        try expectEqual(-std.math.inf(T), gamma(T, -0.0));
+        try expectEqual(std.math.inf(T), gamma(T, 0.0));
+
+        try expectEqual(@as(T, -0.0), gamma(T, -200.5));
+        try expectEqual(@as(T, 0.0), gamma(T, -201.5));
+        try expectEqual(@as(T, -0.0), gamma(T, -202.5));
+
+        try expectEqual(std.math.inf(T), gamma(T, 200));
+        try expectEqual(std.math.inf(T), gamma(T, 201));
+        try expectEqual(std.math.inf(T), gamma(T, 202));
+
+        try expectEqual(std.math.inf(T), gamma(T, std.math.inf(T)));
+    }
+}
+
+test "math.lgamma" {
+    inline for (&.{f32, f64}) |T| {
+        const eps = @sqrt(std.math.floatEps(T));
```

Comment:
1. This looks confusing to me. Why do you assign `eps = sqrt_eps` here ? 
2. What does the standard mandate about precision of lgamma?

Reply:
`std.math.approxEqRel()` recommends using `sqrt(floatEps(T))` so I used it, in practice tests passed on my machine for an `eps` 100 times smaller, but i kept this as precision can depend on the architecture but most likely never worse than `sqrt(floatEps(T))`.

Should I change the name of the `eps` variable to reflect that?

For comparison, `std.math.pow()` uses `const epsilon = 0.000001;`.

Reply:
From what i understand, C standard doesn't mandate any precision on math functions using floats.

Reply:
> From what i understand, C standard doesn't mandate any precision on math functions using floats.

My bad actually. I was wondering, what musl tells about the accuracy/precision. The mandated precision of math functions should be specified in IEEE754, but so far I have not read the what the spec says must be precision in ULP for each function. 
You can read an overview of accuracy in the paper "Accuracy of Mathematical Functions in Single, Double, and Quadruple Precision" by Paul Zimmermann.

It would be nice to have the exact numbers. Generally, so far (and if I remember correctly, I have been using for f16 0.001, f32 10^-6, f64 10^-12 and f128 10^-24.
However, it would be of course nice, if we could validate that this is correct to comply with IEEE754 and ideally provide a better lower bound.
The first step would be to have the overview for math function accuracies for f16, f32, f64, f128.

This would be no strict requirement, but a strict one for eventual compiler_rt validation.

Reply:
musl doesn't tell it, python implementation say it's less than 10 ULP
https://github.com/python/cpython/blob/50b093f5c7060c0b44c264808411346cee7becf0/Modules/mathmodule.c#L271